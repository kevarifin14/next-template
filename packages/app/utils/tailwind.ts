export const classNames = (...classes) => classes.filter(Boolean).join(' ');

export const sizeToDimension = {
  xs: 6,
  sm: 8,
  md: 10,
  lg: 12,
  xl: 16,
  '2xl': 20,
  '3xl': 24,
  '4xl': 28,
};
