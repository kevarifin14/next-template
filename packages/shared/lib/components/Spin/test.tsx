import { render } from '@testing-library/react';
import React from 'react';

import { Spin } from '.';

describe('Spin', () => {
  it('matches snapshot', () => {
    const { baseElement } = render(<Spin />);
    expect(baseElement).toMatchSnapshot();
  });
});
