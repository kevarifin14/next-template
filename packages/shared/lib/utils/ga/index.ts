export const pageview = (url) => {
  window.gtag("config", process.env.NEXT_PUBLIC_GOOGLE_ANALYTICS, {
    page_path: url,
  });
};

type EventCategory =
  | "lead"
  | "outbound link"
  | "call to action"
  | "login"
  | "swap";

export const event = (
  category: EventCategory,
  action,
  label = "",
  params = {}
) => {
  if (process.env.NODE_ENV === "production") {
    window.gtag("event", action, {
      event_category: category,
      event_label: label,
      ...params,
    });
  }
};
