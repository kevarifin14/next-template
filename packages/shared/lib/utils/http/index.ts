import QueryString from 'query-string';

export function generateQueryString(params) {
  return QueryString.stringify(params);
}

export function generateEndpoint(endpoint, params) {
  if (Object.keys(params).length === 0) {
    return endpoint;
  }
  return `${endpoint}?${generateQueryString(params)}`;
}

export const get = async (endpoint, params = {}) => {
  const endpointWithParams = generateEndpoint(endpoint, params);
  const response = await fetch(endpointWithParams);

  if (response.status === 200) {
    const responseJson = response.json();
    return responseJson;
  }
  throw new Error(`${response.status}`);
};

export const post = async (endpoint, body, headers?) => {
  const response = await fetch(
    endpoint,
    {
      method: 'POST',
      body: JSON.stringify(body),
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
    },
  );

  if (response.status >= 200 && response.status < 400) {
    const responseJson = response.json();
    return responseJson;
  }
  throw new Error(`${response.status}`);
};
