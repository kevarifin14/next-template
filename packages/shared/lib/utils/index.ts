export function splitName(name: string) {
  return {
    firstName: name.split(" ").slice(0, -1).join(" "),
    lastName: name.split(" ").slice(-1).join(" "),
  };
}
