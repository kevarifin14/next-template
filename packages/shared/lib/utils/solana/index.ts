/* istanbul ignore file */
import {
  AccountLayout,
  u64,
  AccountInfo as TokenAccountInfo,
  MintInfo,
} from "@solana/spl-token";
import { PublicKey } from "@solana/web3.js";

export const deserializeTokenAccount = (address: PublicKey, data: Buffer) => {
  const {
    mint,
    owner,
    amount,
    delegateOption,
    delegate,
    delegatedAmount,
    state,
    isNative,
    isNativeOption,
    closeAuthority,
    closeAuthorityOption,
  } = AccountLayout.decode(data);

  const accountInfo: TokenAccountInfo = {
    address,
    mint: new PublicKey(mint),
    owner: new PublicKey(owner),
    amount: u64.fromBuffer(amount),
    isInitialized: state !== 0,
    isFrozen: state === 2,

    ...(delegateOption === 0
      ? { delegate: null, delegatedAmount: new u64(0) }
      : {
          delegate: new PublicKey(delegate),
          delegatedAmount: u64.fromBuffer(delegatedAmount),
        }),

    ...(isNativeOption === 1
      ? { rentExemptReserve: u64.fromBuffer(isNative), isNative: true }
      : { rentExemptReserve: null, isNative: false }),

    ...(closeAuthorityOption === 0
      ? { closeAuthority: null }
      : { closeAuthority: new PublicKey(closeAuthority) }),
  };

  return accountInfo;
};

export const diffTokenAccountInfoAmounts = (
  prevTokenAccountInfo: TokenAccountInfo,
  tokenAccountInfo: TokenAccountInfo
) => {
  if (!prevTokenAccountInfo) {
    return tokenAccountInfo.amount;
  }

  return tokenAccountInfo.amount.sub(prevTokenAccountInfo.amount);
};

export const fromLamports = (
  tokenAccountInfo: TokenAccountInfo | number,
  mintInfo?: MintInfo
) => {
  if (!tokenAccountInfo) {
    return 0;
  }

  const amount =
    typeof tokenAccountInfo === "number"
      ? tokenAccountInfo
      : tokenAccountInfo.amount.toNumber();

  const precision = 10 ** (mintInfo?.decimals || 9);
  return amount / precision;
};

export function formatTokenAmount(
  tokenAccountInfo: TokenAccountInfo | number,
  mintInfo?: MintInfo,
  prefix = "",
  suffix = "",
  precision = 3
): string {
  return `${[prefix]}${fromLamports(tokenAccountInfo, mintInfo).toFixed(
    precision
  )}${suffix}`;
}

export const formatDollar = (value) => {
  const formatter = new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
  });

  return formatter.format(value);
};
