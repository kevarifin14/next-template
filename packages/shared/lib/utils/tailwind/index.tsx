import { ComponentProps } from 'react';

export type TailwindSize = 'xs' | 'sm' | 'md' | 'lg' | 'xl' | '2xl' | '3xl' | '4xl';

export const classNames = (...classes) => classes.filter(Boolean).join(' ');

export const sizeToDimension = {
  xs: 6,
  sm: 8,
  md: 10,
  lg: 12,
  xl: 16,
  '2xl': 20,
  '3xl': 24,
  '4xl': 28,
};

export const sizeToClassName = (size: TailwindSize) => {
  const dimension = sizeToDimension[size];
  return `h-${dimension} w-${dimension}`;
};

export type AlertType = 'success' | 'info' | 'error' | 'warning';

export type Heroicon = (props: ComponentProps<'svg'>) => JSX.Element; // eslint-disable-line