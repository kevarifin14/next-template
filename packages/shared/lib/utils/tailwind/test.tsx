import { sizeToClassName } from '.';

describe('tailwind', () => {
  describe('sizeToClassName', () => {
    it('converts size to className dimensions', () => {
      expect(sizeToClassName('xs')).toBe('h-6 w-6');
    });
  });
});
