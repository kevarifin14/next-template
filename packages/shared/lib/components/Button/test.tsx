import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';

import { Button } from '.';
import { classNames } from '../../utils/tailwind';
import * as SpinComponent from '../Spin';

describe('Button', () => {
  const mockChildren = 'Click';

  const spinSpy = jest.spyOn(SpinComponent, 'Spin');

  const renderButton = (props) => {
    render(<Button {...props}>{mockChildren}</Button>);
    return screen.getByRole('button');
  };

  afterEach(() => jest.clearAllMocks());

  it('matches snapshot', () => {
    const { baseElement } = render(
      <Button type="primary">{mockChildren}</Button>,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('renders the child', () => {
    const button = renderButton({});
    expect(button).toContainHTML(mockChildren);
    expect(spinSpy).not.toHaveBeenCalled();
  });

  describe('Button size', () => {
    it('creates a default sized button', () => {
      const button = renderButton({});
      expect(button).toHaveClass('px-4 py-2');
    });

    it('makes the button lg', () => {
      const button = renderButton({ size: 'lg' });
      expect(button).toHaveClass('px-5 py-3 text-lg');
    });

    it('makes the button xl', () => {
      const button = renderButton({ size: 'xl' });
      expect(button).toHaveClass('px-6 py-4 text-xl');
    });
  });

  it('fires onClick function', () => {
    const mockOnClick = jest.fn();
    const button = renderButton({ onClick: mockOnClick });
    fireEvent.click(button);
    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  it('makes the button primary', () => {
    const button = renderButton({ type: 'primary' });
    expect(button).toHaveClass('bg-primary');
  });

  it('makes the button primary gradient', () => {
    const button = renderButton({ type: 'primary', gradient: true });
    expect(button).toHaveClass('bg-gradient-to-tr from-primary-500 to-primary-900 text-white');
  });

  it('makes the button secondary', () => {
    const button = renderButton({ type: 'secondary' });
    expect(button).toHaveClass(
      classNames(
        'text-primary-dark bg-primary-100 border-primary-100 hover:bg-primary-200',
        'hover:border-primary-200 border',
        'dark:text-primary dark:bg-transparent dark:border-primary',
      ),
    );
  });

  it('makes the button loading', () => {
    const button = renderButton({ loading: true });

    expect(button).toBeDisabled();
    expect(spinSpy).toHaveBeenCalled();
  });

  it('makes the button disabled', () => {
    const button = renderButton({ disabled: true });

    expect(button).toHaveClass('opacity-50');
    expect(button).toBeDisabled();
    expect(spinSpy).not.toHaveBeenCalled();
  });

  it('makes the button block', () => {
    const button = renderButton({ block: true });

    expect(button).toHaveClass('w-full');
  });
});
