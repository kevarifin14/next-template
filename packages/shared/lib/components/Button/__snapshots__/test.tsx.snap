// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches snapshot 1`] = `
<body>
  <div>
    <button
      class="inline-flex items-center justify-center focus:outline-none font-medium rounded-md text-white bg-primary border-primary hover:bg-primary-dark hover:border-primary-dark border    px-4 py-2 "
      type="button"
    >
      Click
    </button>
  </div>
</body>
`;
